dotCMS/src/main/java/com/dotcms/storage/CompositeStoragePersistenceAPI.java

```diff
@@ -0,0 +1,315 @@
+    }
+
+    private void populateMissStorageChain(final String groupName, final String path,
+                                          final List<StoragePersistenceAPI> missStorageList, final File file) {
+
+        if (!missStorageList.isEmpty() && Objects.nonNull(file)) {
+
+            for(final StoragePersistenceAPI storage : missStorageList) {
+
+                storage.pushFileAsync(groupName, path, file, null);
+            }
+        }
+    }
+
+    @Override
+    public Object pullObject(final String groupName,
+                             final String path,
+                             final ObjectReaderDelegate readerDelegate) throws DotDataException {
+        // todo: here we will look for on each storage the Object, if an upper layer does not have the object, we will look for it in the next one
+        // at the end of the process for each storage without the object, an async call will be execute in order to populate the storage with that particular object
+        Object objectToReturn = null;
+
+        final List<StoragePersistenceAPI> missStorageList = new ArrayList<>();
+        for(final StoragePersistenceAPI storage : this.storagePersistenceAPIList) {
+
+            final Object localObject = storage.pullObject(groupName, path, readerDelegate);
+            if (null != localObject) {
+
+                objectToReturn = localObject;
+                break;
+            } else {
+
+                missStorageList.add(storage); // we will populate this list with the storages that does not have the object, so will be populated async at the end
+            }
+        }
+
+        populateMissStorageChain(groupName, path, missStorageList, objectToReturn);
+
+        return objectToReturn;
+    }
+
+    private void populateMissStorageChain(final String groupName, final String path,
+                                          final List<StoragePersistenceAPI> missStorageList, final Object objectToReturn) {
+
+        if (!missStorageList.isEmpty() && Objects.nonNull(objectToReturn) && objectToReturn instanceof Serializable) {
+
+            for(final StoragePersistenceAPI storage : missStorageList) {
+
+                storage.pushObjectAsync(groupName, path, defaultWriterDelegate, Serializable.class.cast(objectToReturn), null);
+            }
+        }
+    }
+
+    @Override
+    public Future<File> pullFileAsync(final String groupName, final String path) {
+
+        final List<Future<File>> futures = new ArrayList<>();
+        for(final StoragePersistenceAPI storage : this.storagePersistenceAPIList) {
+
+            final Future<File> localFileFuture = storage.pullFileAsync(groupName, path);
+            futures.add(localFileFuture);
+        }
+
+        // since we do not want to wait for all the futures to be completed, we do not which layer does not have the file, so we fire a new thread to call the sync method
+        DotConcurrentFactory.getInstance().getSubmitter().submit(()-> this.pullFile(groupName, path));
```

Comment:
Should this be its own submitter group for greater configurability?

Reply:
ok sounds fair